### PR TITLE
init unlocker

### PR DESCRIPTION
**Pull request by franzns**

### CODE REVIEW

File: contracts/VaultUnlocker.sol [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        // calculate the penalty
+        uint256 totalFtmAfter = _staking.totalFTMWorth();
+        uint256 penalty = totalFtmBefore - totalFtmAfter;
+        require(penalty <= msg.value, "ERR_INSUFFICIENT_FUNDS");
+
+        // send FTM to the vault to make up for penalty
+        address stakingAddress = address(_staking);
+        (bool sentToStaking, bytes memory dataStaking) = stakingAddress.call{value: penalty}("");
+        require(sentToStaking, "Failed to send FTM to staking");
```
Comment (danielmkm) [inline contracts/VaultUnlocker.sol:37]: I'd add more checks than just this:
- Get the FTM balance before and after, make sure `difference == penalty`
- ensure `_staking.totalFTMWorth() == totalFtmBefore`, if not revert

File: contracts/VaultUnlocker.sol [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        uint256 totalFtmBefore = _staking.totalFTMWorth();
+
+        // unlocking the vault, this will pay the penalty and not the complete stake will be returned
+        uint256 currentStake = _vault.currentStakeValue();
+        _vault.unlock(currentStake);
+
+        // calculate the penalty
+        uint256 totalFtmAfter = _staking.totalFTMWorth();
+        uint256 penalty = totalFtmBefore - totalFtmAfter;
+        require(penalty <= msg.value, "ERR_INSUFFICIENT_FUNDS");
```
Comment (danielmkm) [inline contracts/VaultUnlocker.sol:32]: does this need to be `<` or `<=`

File: contracts/VaultUnlocker.sol [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    address public immutable _treasury;
+
+    constructor(FTMStaking ftmStaking, Vault vault, address treasury) {
+        _staking = ftmStaking;
+        _vault = vault;
+        _treasury = treasury;
+    }
+
+    function unlockVault() external payable {
+        require(_vault.owner() == address(this), "ERR_UNAUTHORIZED");
```
Comment (danielmkm) [inline contracts/VaultUnlocker.sol:20]: Would not let this be permissionless, regardless if its ok, enforce that `msg.sender` is a known address.

File: contracts/VaultUnlocker.sol [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        uint256 penalty = totalFtmBefore - totalFtmAfter;
+        require(penalty <= msg.value, "ERR_INSUFFICIENT_FUNDS");
+
+        // send FTM to the vault to make up for penalty
+        address stakingAddress = address(_staking);
+        (bool sentToStaking, bytes memory dataStaking) = stakingAddress.call{value: penalty}("");
+        require(sentToStaking, "Failed to send FTM to staking");
+
+        // revert the ownership back to the staking contract
+        _vault.updateOwner(stakingAddress);
```
Comment (danielmkm) [inline contracts/VaultUnlocker.sol:40]: There's no way to get ownership back if something goes wrong, add a permissioned function that does just this.

File: contracts/VaultUnlocker.sol (hunk 2 above)
Comment (franzns) [reply contracts/VaultUnlocker.sol:32]: might be safer to use `<` although I believe `<=` also works

File: contracts/VaultUnlocker.sol [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    Vault public immutable _vault;
+    address public immutable _treasury;
+
+    constructor(FTMStaking ftmStaking, Vault vault, address treasury) {
+        _staking = ftmStaking;
+        _vault = vault;
+        _treasury = treasury;
+    }
+
+    function unlockVault() external payable {
```
Comment (danielmkm) [inline contracts/VaultUnlocker.sol:19]: payable is fine, but i'd probably just send the contract ftm first then call this function so it's a bit easier to manage. Leave it to you though.

File: contracts/VaultUnlocker.sol (hunk 1 above)
Comment (franzns) [reply contracts/VaultUnlocker.sol:38]: > Get the FTM balance before and after, make sure difference == penalty

does that make sense when I already do `uint256 penalty = totalFtmBefore - totalFtmAfter;` ?

added the totalFTMWorth check

**Pull request merged**